src/DurableTask/AsyncPageable{T}.cs

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.DurableTask;
+
+/// <summary>
+/// A collection of values that may take multiple service requests to iterate over.
+/// </summary>
+/// <typeparam name="T">The type of the value.</typeparam>
+public abstract class AsyncPageable<T> : IAsyncEnumerable<T>
+    where T : notnull
+{
+    // This code was adapted from Azure SDK AsyncPageable.
+
+    /// <summary>
+    /// Enumerate the values a <see cref="Page{T}" /> at a time.
+    /// </summary>
+    /// <param name="continuationToken">
+    /// A continuation token indicating where to resume paging or null to begin paging from the
+    /// beginning.
+    /// </param>
+    /// <param name="pageSizeHint">
+    /// The number of items per <see cref="Page{T}"/> that should be requested
+    /// (from service operations that support it). It's not guaranteed that the value will be
+    /// respected.
+    /// </param>
+    /// <returns>An async enumerable of pages.</returns>
+    public abstract IAsyncEnumerable<Page<T>> AsPages(
+        string? continuationToken = default, int? pageSizeHint = default);
+
+    /// <inheritdoc/>
+    public async IAsyncEnumerator<T> GetAsyncEnumerator(CancellationToken cancellationToken = default)
+    {
+        await foreach (Page<T> page in this.AsPages()
+            .ConfigureAwait(false).WithCancellation(cancellationToken))
```

Comment:
Is `.ConfigureAwait(false)` really necessary? It makes me really nervous because of how it can cause code to break in certain contexts.

Reply:
It is safe (and recommended) for libraries to call `ConfigureAwait(false)`. You will continue on the thread pool. but only your continuation is affected. This will not affect the `ConfigureAwait` behavior of the caller to this function - so if they are on a specialized context, their continuation will still be posted back to their special context when our work completes.

Reply:
Can you explain how it would cause code to break in certain contexts? I can only see it breaking if this function itself depended on a specific context. Or it was calling into something that also depended on a specific context. However, this code does not have any of those dependencies.

Reply:
> You will continue on the thread pool, but only your continuation is affected.

DTFx orchestration code is an example where forcing a callback onto the thread pool can break the application. Obviously, we don't expect anyone writing orchestration code to call into these APIs, but there may be other cases where thread pool threads are not desirable.

That said, this might be a complete knowledge gap on my part. When you say, "your continuation", do you mean that the caller of the async enumerator will have its callbacks posted to any context it requires? So as an example, if I (for some strange reason) needed to call this from an Orchestration thread, would the callbacks still be posted back to the orchestration thread instead of a worker pool thread?

Reply:
I actually do see this type as very useful in orchestrations. Someone could write an activity which yields a page at a time, and then use an `AsyncPageable<T>` extension method to create a streamlined experience for calling the activity from an orchestration. How will this `ConfigureAwait(false)` affect that though is a great question. I forgot the detail of DTFx sync context is it based entirely around executing inline. Will need to think about this a bit more.

Reply:
I did some testing (the integration test infra you have works great btw!) and yeah - `ConfigureAwait(false)` breaks using this within an orchestration.

Reply:
I have removed `ConfigureAwait(false)` for now.